---
title: "Fitting parameters"
author: "Mackenzie Johnson"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Varying the number of parameters used for fitting amino acid distributions

This is the RMarkdown document used to produce Figure 6 and Supplementary Figure 6 and 7. It compares results for fitting distributions with a different number of parameters.

Load necessary packages
```{r, message=FALSE}
library(tidyverse)
library(Biostrings)
library(stringr)
library(cowplot)
library(broom)
library(ggtext)
theme_set(theme_cowplot())

```


Read in and format data
```{r, message=FALSE}
# read in all .csv files in /data/real/
empirical_files <- dir(
  "data/real", 
  pattern = "*_Aligned_Sequences.fasta", 
  full.names = T
)

# import alignment
import_alignment <- function(filename) {
  
  fasta_file <- readAAStringSet(filename, format = "fasta")
  seq_name = names(fasta_file)
  aa_seq = paste(fasta_file)
  seq_num = seq(1:length(seq_name))
  real_align <- data.frame(seq_num, aa_seq, stringsAsFactors = F)
  
}

# change sequences from strings into character vectors
aa_string_to_vect <- function(real_align) {

  as.data.frame(strsplit(real_align$aa_seq, split = ""), 
                row.names = NULL) -> int_df
  as.data.frame(t(int_df), row.names = NULL) -> int_df

}

# tidy data frame
tidy_df <- function(int_df) {
  int_df %>%
    mutate(num_seq = seq(1:nrow(int_df))) %>% 
    pivot_longer(
      -num_seq, 
      names_to = "site_raw", 
      values_to = "aa"
    ) -> int_df
  int_df %>% 
    extract(
      site_raw, 
      "site", 
      "V(.*)", 
      remove = TRUE
    ) -> int_df
}

#import_alignment(empirical_files[1]) %>% View()
tibble(filename = empirical_files[1]) %>% 
  extract(filename, "protein", "data/real/(.*)_A", remove = FALSE) %>% 
  mutate(
    raw_data = map(filename, import_alignment),
    long_data = map(raw_data, aa_string_to_vect),
    tidy_data = map(long_data, tidy_df)
  ) %>% 
  select(-filename, -raw_data, -long_data) %>% 
  unnest(cols = tidy_data) -> alignment

# replace undefined aas
alignment$site <- as.numeric(alignment$site)

levels(alignment$aa) # "X" = any aa, "B" = N or D, "Z" = Q or D

alignment$aa <- recode(alignment$aa, X = "-", B = "N", Z = "Q")

levels(alignment$aa)

alignment$aa <- as.character(alignment$aa)

```


Actual distribution
```{r, message=FALSE}
# determine the number of each AA at each site
count_aa <- function(protein_alignment) {
  
  protein_alignment %>%
    group_by(site, aa) %>% 
    summarize(count = n())

  }

alignment %>%
  nest(data = c(num_seq, site, aa)) %>% 
  mutate(counts = map(data, count_aa)) %>%
  select(-data) %>% 
  unnest(cols = counts) -> site_aa_count

empty_counts <- tibble(aa = c('A', 'C', 'D', 'E', 'F', 'G', 'H', 
                              'I', 'K', 'L', 'M', 'N', 'P', 'Q', 
                              'R', 'S', 'T', 'V', 'W', 'Y'), 
                           count = rep(0, 20))


# 1. First write a function that can take a data frame with amino acids at one site, and return a data frame with counts, including zeros.
count_zero_aa_site <- function(site_df) {
  
  site_df %>% 
    group_by(aa) %>% 
    summarize(count = n()) -> counts
  full_join(counts, anti_join(empty_counts, counts, by = 'aa'))

  }

# 2. Function to call that function for each protein.
count_zero_aa_protein <- function(protein_df) {
  
  protein_df %>%
    nest(data = -site) %>%
    mutate(counts = map(data, count_zero_aa_site)) %>%
    select(-data) %>%
    unnest(cols = counts)

}

# 3. Apply to all proteins
alignment %>%
  nest(data = c(num_seq, site, aa)) %>%
  mutate(all_counts = map(data, count_zero_aa_protein)) %>%
  select(-data) %>%
  unnest(cols = all_counts) -> all_count


```


Estimated distribution
```{r, message=FALSE}
# USING ONLY OBSERVED AA
# order aa by relative frequency
order_aa <- function(protein_df) {
  
  protein_df %>%
    group_by(site) %>% 
    arrange(desc(count), .by_group = T)
  
}

all_count %>%
  nest(data = c(site, aa, count)) %>% 
  mutate(ordered_count = map(data, order_aa)) %>%
  select(-data) %>% 
  unnest(cols = ordered_count) -> ordered_count

# EXCLUDE all '-' entries/gaps in alignment
ordered_count %>%
  filter(aa != "-") -> no_gaps

# make the categorical variable (aa) numerical (k) by numbering aa's 1-20 in order of freq
number_aa <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(k = as.numeric(0:19))
  
}

no_gaps %>% 
  nest(data = c(site, aa, count)) %>% 
  mutate(add_k = map(data, number_aa)) %>% 
  select(-data) %>% 
  unnest(cols = add_k) -> ordered_count

# remove aa with zero values
ordered_count %>%
  filter(count > 0) -> observed_aa

# count relative to most frequent - rescale to 1 for highly conserved sites
add_rel_count <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(rel_count = count/max(count))
  
}

observed_aa %>%
  nest(data = c(site, aa, count, k)) %>% 
  mutate(rel_count = map(data, add_rel_count)) %>% 
  select(-data) %>% 
  unnest(cols = rel_count) -> observed_aa

# log-transformed data
add_ln_count <- function(protein_df) {
  
  protein_df %>% 
    mutate(ln_count = log(rel_count))
  
}

observed_aa %>%
  nest(data = c(site, aa, count, k, rel_count)) %>% 
  mutate(log_count = map(data, add_ln_count)) %>% 
  select(-data) %>% 
  unnest(log_count) -> observed_aa


```


Add constant to all
```{r, message=FALSE}
# add a small constant to include aa with zero values
ordered_count %>%
  mutate(count = count + 0.01) -> observed_aa_mod
  #mutate(count = count + 0.000001) -> observed_aa_mod

# count relative to most frequent - rescale to 1 for highly conserved sites
observed_aa_mod %>%
  group_by(site) %>%
  mutate(rel_count = count/max(count)) -> observed_aa_mod

# log-transformed data
observed_aa_mod %>%
  mutate(ln_count = log(rel_count)) -> observed_aa_mod

# get rid of old data sets
rm(alignment, all_count, empty_counts, no_gaps, site_aa_count)

```


Look at how different constants/pseudocounts impact fit
```{r, message=FALSE}
# ordered_count %>%
#   mutate(
#     count1 = count + 0.000001,
#     count2 = count + 0.001,
#     count3 = count + 0.01,
#     count4 = count + 0.1
#   ) -> test_df
# 
# test_df %>%
#   group_by(site) %>%
#   mutate(
#     rel_count1 = count1/max(count1),
#     rel_count2 = count2/max(count2),
#     rel_count3 = count3/max(count3),
#     rel_count4 = count4/max(count4)
#   ) -> test_df
# 
# test_df %>%
#   mutate(
#     ln_count1 = log(rel_count1),
#     ln_count2 = log(rel_count2),
#     ln_count3 = log(rel_count3),
#     ln_count4 = log(rel_count4)
#   ) -> test_df
# 
# test_df %>%
#   filter(site == 1) %>% lm(ln_count1 ~ 0 + k, .)
# test_df %>%
#   filter(site == 1) %>% lm(ln_count2 ~ 0 + k, .)
# test_df %>%
#   filter(site == 1) %>% lm(ln_count3 ~ 0 + k, .)
# test_df %>%
#   filter(site == 1) %>% lm(ln_count4 ~ 0 + k, .)
# 
# test_df %>%
#   filter(site == 1) %>%
#   ggplot(aes(x = k)) +
#   geom_point(aes(y = ln_count1), color = "red") +
#   geom_abline(intercept = 0, slope = -1.293, color = "red") +
#   geom_point(aes(y = ln_count2), color = "black") +
#   geom_abline(intercept = 0, slope = -0.7894, color = "black") +
#   geom_point(aes(y = ln_count3), color = "blue") +
#   geom_abline(intercept = 0, slope = -0.6216, color = "blue") +
#   geom_point(aes(y = ln_count4), color = "green") +
#   geom_abline(intercept = 0, slope = -0.4538, color = "green")
# 
# test_df %>%
#   filter(site == 15) %>% lm(ln_count1 ~ k, .)
# test_df %>%
#   filter(site == 15) %>% lm(ln_count2 ~ k, .)
# test_df %>%
#   filter(site == 15) %>% lm(ln_count3 ~ k, .)
# test_df %>%
#   filter(site == 15) %>% lm(ln_count4 ~ k, .)
# 
# test_df %>%
#   filter(site == 15) %>%
#   ggplot(aes(x = k)) +
#   geom_point(aes(y = ln_count1), color = "red") +
#   geom_abline(intercept = -15.35, slope = -0.27, color = "red") +
#   geom_point(aes(y = ln_count2), color = "black") +
#   geom_abline(intercept = -9.73, slope = -0.17, color = "black") +
#   geom_point(aes(y = ln_count3), color = "blue") +
#   geom_abline(intercept = -7.85, slope = -0.14, color = "blue") +
#   geom_point(aes(y = ln_count4), color = "green") +
#   geom_abline(intercept = -5.98, slope = -0.10, color = "green")
# 
# test_df %>%
#   filter(site == 1) %>% lm(ln_count1 ~ k + I(k^2), .)
# test_df %>%
#   filter(site == 1) %>% lm(ln_count2 ~ k + I(k^2), .)
# test_df %>%
#   filter(site == 1) %>% lm(ln_count3 ~ k + I(k^2), .)
# test_df %>%
#   filter(site == 1) %>% lm(ln_count4 ~ k + I(k^2), .)
# 
# fun.1 <- function(k) 2.3072 - 3.652*k + 0.1228*k*k
# fun.2 <- function(k) 0.98411 - 1.94114*k + 0.07362*k*k
# fun.3 <- function(k) 0.54431 - 1.49991*k + 0.05723*k*k
# fun.4 <- function(k) 0.11556 - 1.05967*k + 0.04085*k*k
# 
# test_df %>%
#   filter(site == 1) %>%
#   ggplot(aes(x = k)) +
#   geom_point(aes(y = ln_count1), color = "red") +
#   stat_function(fun = fun.1, color = "red") +
#   geom_point(aes(y = ln_count2), color = "black") +
#   stat_function(fun = fun.2, color = "black") +
#   geom_point(aes(y = ln_count3), color = "blue") +
#   stat_function(fun = fun.3, color = "blue") +
#   geom_point(aes(y = ln_count4), color = "green") +
#   stat_function(fun = fun.4, color = "green")

```


1 parameter fit - without constant
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# set intercept to 0, only fitting 1 parameter (slope)

fit_lm_protein_1 <- function(protein_df) {
  
  protein_df %>% 
    nest(data = c(aa, count, k, rel_count, ln_count)) %>% 
    mutate(
      fit = map(data, ~ lm(ln_count ~ 0 + k, data = .)),
      slope_1 = map_dbl(fit, ~ (.)$coefficients[1]),
      fit_sum = map(fit, summary),
      r2_1 = map_dbl(fit_sum, ~ (.)$adj.r.squared)
    ) %>% 
    select(-data, -fit, -fit_sum)
  
}

observed_aa %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_1)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_1

```


1 parameter fit - with constant
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# set intercept to 0, only fitting 1 parameter (slope)

observed_aa_mod %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_1)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_mod_1

```


2 parameters fit - without constant
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# fitting 2 parameters: slope and intercept

fit_lm_protein_2 <- function(protein_df) {
  
  protein_df %>% 
    nest(data = c(aa, count, k, rel_count, ln_count)) %>% 
    mutate(
      fit = map(data, ~ lm(ln_count ~ k, data = .)),
      intercept_2 = map_dbl(fit, ~ (.)$coefficients[1]),
      slope_2 = map_dbl(fit, ~ (.)$coefficients[2]),
      fit_sum = map(fit, summary),
      r2_2 = map_dbl(fit_sum, ~ (.)$adj.r.squared)
    ) %>% 
    select(-data, -fit, -fit_sum)
  
}

observed_aa %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_2)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_2

```


2 parameters fit - with constant
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# fitting 2 parameters: slope and intercept

observed_aa_mod %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_2)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_mod_2

```


3 parameters fit - with constant
```{r, message=FALSE}
# fit quadratic model to normalized, log-transformed data
# fitting 3 parameter: slope, slope^2, intercept
# use non-linear model, fit to all aa (add small value to zero counts)

observed_aa_mod %>%
  nest(data = -site) %>%
  mutate(
    fit = map(data, ~ lm(ln_count ~ k + I(k^2), data = .)),
    intercept_3 = map_dbl(fit, ~ (.)$coefficients[1]),
    slope_3 = map_dbl(fit, ~ (.)$coefficients[2]),
    sqr_slp_3 = map_dbl(fit, ~ (.)$coefficients[3]),
    fit_sum = map(fit, summary),
    r2_3 = map_dbl(fit_sum, ~ (.)$adj.r.squared)
  ) %>%
  select(-data, -fit, -fit_sum) -> observed_fits_3


```


3 parameter fit - ridge regression
```{r, message=FALSE}
# use ridge regression instead
#install.packages("glmnet", repos = "https://cran.us.r-project.org")
library(glmnet)

# prep dataframe for use in glmnet
observed_aa_mod %>% 
  mutate(k2 = k*k) -> observed_aa_mod

# find a good lambda value to use
find_ridge_lambda <- function(site_df){
  
  cv.glmnet(
    cbind(site_df$k, site_df$k2),
    site_df$ln_count,
    alpha = 0,    # alpha=0 for ridge regression
    grouped = FALSE # necessary bc of limited # of observations
  ) -> cvfit
  
  cvfit$lambda.min -> ridge_lambda
  
  return(ridge_lambda)
  
}

# get rid of sites with only one unique aa
# cv.glmnet cannot fit these sites
# note: this code sometimes won't fit sites with only 2 unique
# amino acids observed
observed_aa_mod %>% 
  group_by(site) %>% 
  filter(k == 1 & count == 0.01) %>% 
  #filter(k == 1 & count == 0.000001) %>% 
  select(site) -> conserved_sites

observed_aa_mod %>% 
  filter(site %in% conserved_sites$site == FALSE) %>% 
  nest(data = -site) %>% 
  mutate(
    optimal_ridge_lambda = map(data, find_ridge_lambda)
  ) %>% 
  unnest(cols = data) %>% 
  select(site, optimal_ridge_lambda) %>% 
  unique() -> site_cv_lambda

site_cv_lambda$optimal_ridge_lambda <- as.numeric(site_cv_lambda$optimal_ridge_lambda)

# # lambda value across sites - use one lambda values (0.65) 0.375 for 0.001 constant
# ggplot(
#   site_cv_lambda,
#   aes(x = optimal_ridge_lambda)
# ) + geom_density() +
#   geom_vline(xintercept = 0.28)

summary(site_cv_lambda$optimal_ridge_lambda)

# note: https://glmnet.stanford.edu/articles/glmnet.html#linear-regression
# there is some discussion in the above reference that predict() should be
# used over fit() after cross validation is performed; however, since we use
# cv.glmnet() to find one value of lambda across sites and are not reusing the 
# same model at each site as we did in cv, we perform a new fit() with set 
# lambda and alpha

fit_ridge <- function(site_df){
  
  # fit model with specific alpha and lambda
  glmnet(
    cbind(site_df$k, site_df$k2),
    site_df$ln_count,
    alpha = 0, # for ridge regression
    lambda = 0.28 # optimal across sites
    #lambda = 0 # used to confirm that you can recreate 3 param model
  ) -> fit
  
  # get coefficients from fit
  coef(fit) -> fit_params
  
  # return coefficients
  return(fit_params)
  
}

# apply to all sites
observed_aa_mod %>%
  nest(data = -site) %>%
  mutate(
    fit = map(data, fit_ridge),
    intercept_ridge = map_dbl(fit, ~ (.)[1]),
    slope_ridge = map_dbl(fit, ~ (.)[2]),
    sqr_slp_ridge = map_dbl(fit, ~ (.)[3])
  ) %>%
  select(-data, -fit) -> observed_fits_3_ridge

```


Examine what value of lambda in `glmnet` give you the results of `lm`
```{r, message=FALSE}
# # some lambda value ridge regression should recreate what you get from lm
# observed_aa_mod %>%
#   filter(site == 3) -> one_site_df
# 
# # fits from 3 parameter lm() fit
# # intercept: -7.062859594
# # k coefficient: -2.28672487
# # k squared coefficient: 0.095697526
# 
# # optimal lambda from this site: 0.2701341
# 
# # fits from 3 parameter glmnet() fit with lambda = 0.65
# # intercept: -12.52286734
# # k coefficient: -0.64103414
# # k squared coefficient: 0.013316520
# 
# glmnet(
#     cbind(one_site_df$k, one_site_df$k2),
#     one_site_df$ln_count,
#     alpha = 0 # for ridge regression
#   ) -> fit
# 
# coef(fit) -> fits
# # %Dev 43.51 for lambda = 0.27
# # %Dev ~34 for lambda = 0.65
# 
# coef(fit, s = 0.270)
# 
# glmnet(
#   cbind(one_site_df$k, one_site_df$k2),
#   one_site_df$ln_count,
#   alpha = 0,
#   lambda = 0.0000002
# ) %>% coef()

```


Get estimate counts 
```{r, message=FALSE}
# combine dataframes of fits (no constant for 1 and 2 parameter)
observed_fits_1 %>%
  full_join(., observed_fits_2, by = c("protein", "site")) %>% 
  full_join(., observed_fits_3, by = "site") -> observed_fits

# combine dataframes of fits (add constant to all)
observed_fits_mod_1 %>%
  full_join(., observed_fits_mod_2, by = c("protein", "site")) %>% 
  full_join(., observed_fits_3_ridge, by = "site") -> observed_fits_mod

# extract r2 results
observed_fits %>% 
  select(protein, site, r2_1, r2_2, r2_3) -> regression_performance

# get rid of old subsets
rm(observed_fits_1, observed_fits_2, observed_fits_mod_1, 
   observed_fits_mod_2, observed_fits_3, observed_fits_3_ridge,
   conserved_sites)

# added rel_count and ln_count to all aa
ordered_count %>%
  nest(data = c(site, aa, count, k)) %>% 
  mutate(rel_count = map(data, add_rel_count)) %>% 
  select(-data) %>% 
  unnest(cols = rel_count) -> ordered_count

ordered_count %>%
  nest(data = c(site, aa, count, k, rel_count)) %>% 
  mutate(ln_count = map(data, add_ln_count)) %>% 
  select(-data) %>% 
  unnest(cols = ln_count) -> ordered_count

# estimate counts with parameters from fit - no constant for 1 and 2 param
ordered_count %>% 
  left_join(observed_fits, by = c("protein", "site")) %>% 
  mutate(
    est_dist_1 = slope_1*exp(k*slope_1),
    est_dist_2 = slope_2*exp(k*slope_2+intercept_2),
    est_dist_3 = slope_3*exp(k*slope_3 + k*k*sqr_slp_3 + 
                                    intercept_3)
  ) -> ordered_count

# rename prior to join
colnames(observed_fits_mod)[3] <- "slope_1_mod"
colnames(observed_fits_mod)[4] <- "r2_1_mod"
colnames(observed_fits_mod)[5] <- "intercept_2_mod"
colnames(observed_fits_mod)[6] <- "slope_2_mod"
colnames(observed_fits_mod)[7] <- "r2_2_mod"

# estimate counts with parameters from fit - constant for all
ordered_count %>% 
  left_join(
    observed_fits_mod, 
    by = c("protein", "site")
  ) %>%
  mutate(
    est_dist_1_mod = slope_1_mod*exp(k*slope_1_mod),
    est_dist_2_mod = slope_2_mod*exp(k*slope_2_mod+intercept_2_mod),
    est_dist_3_rid = slope_ridge*exp(
      k*slope_ridge + k*k*sqr_slp_ridge + intercept_ridge
    )
  ) -> ordered_count

# rescale to compare w raw count
rescale_est_dist <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(
      est_rel_1 = est_dist_1/sum(est_dist_1),
      est_rel_2 = est_dist_2/sum(est_dist_2),
      est_rel_3 = est_dist_3/sum(est_dist_3),
      est_rel_1_mod = est_dist_1_mod/sum(est_dist_1_mod),
      est_rel_2_mod = est_dist_2_mod/sum(est_dist_2_mod),
      est_rel_3_rid = est_dist_3_rid/sum(est_dist_3_rid)
    ) %>% 
    mutate(
      est_count_1 = sum(count)*est_rel_1,
      est_count_2 = sum(count)*est_rel_2,
      est_count_3 = sum(count)*est_rel_3,
      est_count_1_mod = sum(count)*est_rel_1_mod,
      est_count_2_mod = sum(count)*est_rel_2_mod,
      est_count_3_rid = sum(count)*est_rel_3_rid
    )
  
}

ordered_count %>%
  nest(data = c(site, aa, count, k, rel_count, 
                ln_count, slope_1, intercept_2,
                slope_2, intercept_3, slope_3,
                sqr_slp_3, intercept_ridge,
                slope_ridge, sqr_slp_ridge,
                est_dist_1, est_dist_2, 
                est_dist_3, est_dist_1_mod, 
                est_dist_2_mod, est_dist_3_rid)) %>% 
  mutate(rescaled = map(data, rescale_est_dist)) %>% 
  select(-data) %>% 
  unnest(cols = rescaled) -> ordered_count

# remove duplicate information
rm(observed_fits, observed_fits_mod)

```


Visualize distributions for debugging
```{r}
# ordered_count %>% 
#   select(site, intercept_3, intercept_ridge, slope_3, slope_ridge, sqr_slp_3, sqr_slp_ridge) %>% 
#   unique() %>% 
#   filter( 
#     intercept_3 > 0 & intercept_ridge < 0 |
#       intercept_3 < 0 & intercept_ridge > 0 |
#       slope_3 > 0 & slope_ridge < 0 |
#       slope_3 < 0 & slope_ridge > 0 |
#       sqr_slp_3 > 0 & sqr_slp_ridge < 0 |
#       sqr_slp_3 < 0 & sqr_slp_ridge > 0
#   )%>% View()

#fun.norm <- function(k) 0.055728816 - 0.650043121*k + 0.0066499756*k*k # site 16
#fun.rid <- function(k) -1.021969165 - 0.3591600*k - 0.0069993493*k*k # site 16
#fun.norm <- function(k) -4.71 - 0.989*k + 0.0417*k*k # site 44
#fun.rid <- function(k) -7.06 - 0.279*k + 0.00610*k*k # site 44
#fun.norm <- function(k) -0.8082027 + 0.063021552*k + -0.02558368*k*k # site 24
#fun.rid <- function(k) -0.2963912 - 0.1385702*k - 0.01422084*k*k # site 24
#fun.norm <- function(k) 0.039615500 - 1.1014443*k + 0.0320175777*k*k # site 2
#fun.rid <- function(k) -2.135163818 - 2.135163818*k + 0.0008326154*k*k # site 2

# fun.3 <- function(k) 0.2896188 - 1.179601*k + 0.04481159*k*k # site 153
# fun.rid <- function(k) 0.2837349 - 1.177892*k + 0.04472777*k*k # site 153
# 
# ordered_count %>%
#   filter(site == 153) %>% # sites to look at: 16, 44, 80, 138
#   ggplot(aes(x = k)) +
#   #geom_point(aes(y = ln_count)) +
#   # geom_path(aes(y = est_dist_1), color = "black") +
#   # geom_path(aes(y = est_dist_2), color = "blue") +
#   #geom_path(aes(y = est_dist_3), color = "grey") 
#   # geom_path(aes(y = est_dist_1_mod), color = "red") +
#   # geom_path(aes(y = est_dist_2_mod), color = "green") +
#   #geom_path(aes(y = est_dist_3_rid), color = "orange") +
#   #stat_function(fun = fun.3, color = "purple") +
#   #stat_function(fun = fun.rid, color = "yellow")
#   # geom_path(aes(y = est_rel_1), color = "black") +
#   # geom_path(aes(y = est_rel_2), color = "blue") +
#   # geom_path(aes(y = est_rel_3), color = "grey") +
#   # geom_path(aes(y = est_rel_1_mod), color = "red") +
#   # geom_path(aes(y = est_rel_2_mod), color = "green") +
#   # geom_path(aes(y = est_rel_3_rid), color = "orange")
#   geom_point(aes(y = count)) +
#   geom_path(aes(y = est_count_1), color = "black") +
#   geom_path(aes(y = est_count_2), color = "blue") +
#   geom_path(aes(y = est_count_3), color = "grey") +
#   geom_path(aes(y = est_count_1_mod), color = "red") +
#   geom_path(aes(y = est_count_2_mod), color = "green") +
#   geom_path(aes(y = est_count_3_rid), color = "orange")


```




Chi-squared test: actual vs estimated
```{r, message=FALSE}
# use raw count
ordered_count %>% 
  nest(data = -site) %>%
  ungroup() %>%
  mutate(
    chisq_1 = map(data, ~ sum(((.$count - .$est_count_1)^2)/.$est_count_1)),
    chisq_2 = map(data, ~ sum(((.$count - .$est_count_2)^2)/.$est_count_2)),
    chisq_3 = map(data, ~ sum(((.$count - .$est_count_3)^2)/.$est_count_3)),
    chisq_1_mod = map(data, ~ sum(((.$count - .$est_count_1_mod)^2)/.$est_count_1_mod)),
    chisq_2_mod = map(data, ~ sum(((.$count - .$est_count_2_mod)^2)/.$est_count_2_mod)),
    chisq_3_rid = map(data, ~ sum(((.$count - .$est_count_3_rid)^2)/.$est_count_3_rid)),
    p_value_1 = map_dbl(chisq_1, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_2 = map_dbl(chisq_2, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_3 = map_dbl(chisq_3, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_1_mod = map_dbl(chisq_1_mod, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_2_mod = map_dbl(chisq_2_mod, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_3_rid = map_dbl(chisq_3_rid, ~ pchisq(., 18, lower.tail=FALSE)),
  ) %>% 
  select(-data) %>%
  mutate(
    p.adjusted_1 = p.adjust(p_value_1, method= "fdr"),
    p.adjusted_2 = p.adjust(p_value_2, method= "fdr"),
    p.adjusted_3 = p.adjust(p_value_3, method= "fdr"),
    p.adjusted_1_mod = p.adjust(p_value_1_mod, method= "fdr"),
    p.adjusted_2_mod = p.adjust(p_value_2_mod, method= "fdr"),
    p.adjusted_3_rid = p.adjust(p_value_3_rid, method= "fdr")
  ) -> chisq_results

chisq_results %>% 
  select(
    site, p.adjusted_1, p.adjusted_2, p.adjusted_3,
    p.adjusted_1_mod, p.adjusted_2_mod, p.adjusted_3_rid
  ) -> chisq_results

pivot_longer(
  chisq_results, 
  -site, 
  names_to = "parameters_raw", 
  values_to = "p.adjusted"
) %>% 
  extract(
    parameters_raw, 
    "parameters", 
    "p.adjusted_(.*)", 
    remove = T
  ) -> chisq_results

chisq_results %>%
  mutate(result = "pass") -> chisq_results

# REMOVE conserved sites with a single aa
chisq_results %>% 
  filter(p.adjusted != "NaN") -> chisq_results

chisq_results$result[chisq_results$p.adjusted < 0.05] <- "fail"

```


Chi-squared results based on number of parameters used
```{r, message=FALSE}
# 133 sites; 17 fail (12.8%) - NO CONSTANT
chisq_results %>% 
  filter(parameters == 1 & result == "fail") %>% nrow()

# 153 sites; 98 fail (64.1%) - WITH CONSTANT
chisq_results %>% 
  filter(parameters == "1_mod" & result == "fail") %>% nrow()

# 133 sites; 34 fail (25.6%) - NO CONSTANT
chisq_results %>% 
  filter(parameters == 2 & result == "fail") %>% nrow()

# 153 sites; 125 fail (81.7%) - WITH CONSTANT
chisq_results %>% 
  filter(parameters == "2_mod" & result == "fail") %>% nrow()

# 153 sites; 110 fail (71.9%) - WITH CONSTANT
chisq_results %>% 
  filter(parameters == 3 & result == "fail") %>% nrow()

# 153 sites; 105 fail (68.6%) - RIDGE
chisq_results %>% 
  filter(parameters == "3_rid" & result == "fail") %>% nrow()

```


Effective number of amino acids
```{r, message=FALSE}
# eff aa for actual distribution
observed_aa %>%
  group_by(site) %>%
  mutate(frequency = (count)/sum(count)) -> site_aa_freq

site_aa_freq %>% 
  mutate(flnf = frequency*log(frequency)) %>%
  group_by(site) %>% 
  summarize(
    entropy = -sum(flnf),
    eff_aa = exp(entropy),
    n = length(unique(aa))
  ) -> eff_aa_all

# eff aa for estimated distributions
ordered_count %>% 
  group_by(site) %>%
  mutate(
    freq_1 = (est_count_1)/sum(est_count_1),
    freq_2 = (est_count_2)/sum(est_count_2),
    freq_3 = (est_count_3)/sum(est_count_3),
    freq_1_mod = (est_count_1_mod)/sum(est_count_1_mod),
    freq_2_mod = (est_count_2_mod)/sum(est_count_2_mod),
    freq_3_rid = (est_count_3_rid)/sum(est_count_3_rid)
  ) %>% 
  select(
    site, aa, k, freq_1, freq_2, freq_3,
    freq_1_mod, freq_2_mod, freq_3_rid
  ) -> site_aa_freq_2

site_aa_freq_2 %>% 
  mutate(
    flnf_1 = freq_1*log(freq_1),
    flnf_2 = freq_2*log(freq_2),
    flnf_3 = freq_3*log(freq_3),
    flnf_1_mod = freq_1_mod*log(freq_1_mod),
    flnf_2_mod = freq_2_mod*log(freq_2_mod),
    flnf_3_rid = freq_3_rid*log(freq_3_rid)
  ) %>% 
  group_by(site) %>%
  summarize(
    entr_1 = -sum(flnf_1),
    entr_2 = -sum(flnf_2),
    entr_3 = -sum(flnf_3),
    entr_1_mod = -sum(flnf_1_mod),
    entr_2_mod = -sum(flnf_2_mod),
    entr_3_rid = -sum(flnf_3_rid),
    eff_aa_est_1 = exp(entr_1),
    eff_aa_est_2 = exp(entr_2),
    eff_aa_est_3 = exp(entr_3),
    eff_aa_est_1_mod = exp(entr_1_mod),
    eff_aa_est_2_mod = exp(entr_2_mod),
    eff_aa_est_3_rid = exp(entr_3_rid)
  ) -> eff_aa_all_est

eff_aa_all_est %>% 
  pivot_longer(
    c(entr_1, entr_2, entr_3, entr_1_mod, entr_2_mod, entr_3_rid), 
    names_to = "parameters",
    values_to = "entropy_est"
  ) %>% 
  extract(
    parameters, 
    "parameters", 
    "entr_(.*)", 
    remove = T
  ) %>% 
  select(site, parameters, entropy_est) -> sub1
  
eff_aa_all_est %>% 
  pivot_longer(
    c(
      eff_aa_est_1, eff_aa_est_2, eff_aa_est_3, 
      eff_aa_est_1_mod, eff_aa_est_2_mod, 
      eff_aa_est_3_rid
    ),
    names_to = "parameter",
    values_to = "eff_aa_est"
  ) %>% 
  extract(
    parameter, 
    "parameters", 
    "eff_aa_est_(.*)", 
    remove = T
  ) %>%
  select(site, parameters, eff_aa_est) -> sub2

full_join(sub1, sub2, by = c("site", "parameters")) -> eff_aa_est

left_join(eff_aa_all, eff_aa_est, by = "site") -> compare_eff_aa

rm(sub1, sub2, site_aa_freq, site_aa_freq_2,
   eff_aa_all, eff_aa_all_est, eff_aa_est)

```


Supplementary Figure 7 - Relationship between regularization $\lambda$ and $n_{eff}$
```{r, message=FALSE}
# distribution of optimal lambda
ggplot(
  site_cv_lambda, 
  aes(x = optimal_ridge_lambda)
) + geom_density(fill = "grey90") +
  geom_vline(xintercept = 0.28, linetype = "dashed") +
  labs(
    x = "Optimal tuning parameter", 
    y = "Density"
  ) + scale_x_continuous(
    limits = c(0, 1),
    expand = c(0, 0)
  ) + scale_y_continuous(
    limits = c(0, 8),
    expand = c(0, 0)
  ) -> figs7_1
  

# relationship between optimal lambda and neff
compare_eff_aa %>% 
  select(site, eff_aa) %>% 
  unique() %>% 
  left_join(., site_cv_lambda) %>% 
  ggplot(aes(
    x = eff_aa, 
    y = optimal_ridge_lambda
  )) +
  geom_point() +
  geom_smooth() +
  geom_hline(aes(yintercept = 0.28), linetype = "dashed") +
  scale_x_continuous(
    limits = c(1, 12.5),
    expand = c(0.015, 0.015)
  ) +
  scale_y_continuous(
    limits = c(0, 0.75),
    expand = c(0, 0)
  ) +
  labs(
    x = "*n*<sub>eff</sub>", 
    y = "Optimal tuning parameter"
  ) +
  theme(
    axis.title.x = element_markdown(),
    axis.title.y = element_markdown()
  ) +
  background_grid() -> figs7_2

plot_grid(figs7_1, figs7_2, ncol = 2,
          align = "h", axis = "bt", 
          rel_widths = c(1, 1.3),
          labels = c("A", "B")) -> figs7

figs7

save_plot("figures7.png", figs7, ncol = 1, nrow = 1, base_height = 5,
          base_asp = 1.618, base_width = NULL)

```



Find $R^2$ values for each parameter
```{r, message=FALSE}

rsq <- function (x, y) { cor(x, y) ^ 2 }

compare_eff_aa %>% 
  filter(parameters == 1 & !is.na(eff_aa_est)) -> compare_p1

rsq(compare_p1$eff_aa, compare_p1$eff_aa_est) # 0.9706805 - NO CONSTANT 

compare_eff_aa %>% 
  filter(parameters == "1_mod" & !is.na(eff_aa_est)) -> compare_p1_mod

rsq(compare_p1_mod$eff_aa, compare_p1_mod$eff_aa_est) # 0.8859595 - WITH CONSTANT

compare_eff_aa %>% 
  filter(parameters == 2 & !is.na(eff_aa_est)) -> compare_p2

rsq(compare_p2$eff_aa, compare_p2$eff_aa_est) # 0.8589123 - NO CONSTANT

compare_eff_aa %>% 
  filter(parameters == "2_mod" & !is.na(eff_aa_est)) -> compare_p2_mod

rsq(compare_p2_mod$eff_aa, compare_p2_mod$eff_aa_est) # 0.2922884 - WITH CONSTANT

compare_eff_aa %>% 
  filter(parameters == 3 & !is.na(eff_aa_est)) -> compare_p3

rsq(compare_p3$eff_aa, compare_p3$eff_aa_est) # 0.4008924 - WITH CONSTANT

compare_eff_aa %>% 
  filter(parameters == "3_rid" & !is.na(eff_aa_est)) -> compare_p3_rid

rsq(compare_p3_rid$eff_aa, compare_p3_rid$eff_aa_est) # 0.06953798 - RIDGE w lambda = 0

rm(compare_p1, compare_p1_mod, compare_p2, compare_p2_mod, compare_p3)

```


Figure 6
```{r, message=FALSE}

signif(0.9706805, digits = 3)
signif(0.8589123, digits = 3)
signif(0.718942, digits = 3)

dat_text <- data.frame(
  label = c("*R*<sup>2</sup> = 0.971", 
            "*R*<sup>2</sup> = 0.859", 
            "*R*<sup>2</sup> = 0.401"), # 0.719 with constant 0.000001
  parameters = c(1, 2, 3),
  x = c(11.25, 11.25, 11.25),
  y = c(2, 2, 2)
)


left_join(compare_eff_aa, chisq_results) %>% 
  filter(!is.na(result)) -> all_results

all_results %>% 
  filter(
    parameters == 1 | parameters == 2 | parameters == 3
  ) -> all_results_original

ggplot(all_results_original, 
       aes(x=eff_aa, y=eff_aa_est, color = result)) + 
  geom_point() +
  geom_abline(slope = 1, intercept = 0) + 
  labs(
    x = "*n*<sub>eff</sub> of actual distribution", 
    y = "*n*<sub>eff</sub> of fit distribution",
    color = "Result"
  ) +
  scale_y_continuous(
    limits = c(0, 13),
    breaks = c(0, 2.5, 5, 7.5, 10.0, 12.5)
  ) +
  scale_color_manual(
    values = c("#E69F00", "#0072B2")
  ) +
  #geom_richtext(data = dat_text, aes(label = label)) #+
  geom_richtext(
    data = dat_text, 
    aes(x = x, y = y,label = label),
    label.color = NA,
    inherit.aes = FALSE
  ) +
  facet_grid(vars(parameters)) + 
  theme_cowplot(12, rel_small = 1) +
  panel_border() +
  background_grid() +
  theme(
    axis.title.x = element_markdown(),
    axis.title.y = element_markdown(),
    strip.background = element_rect(fill = "gray90")) -> fig6

fig6

save_plot("figure6.png", fig6, ncol = 1, nrow = 1, base_height = 5,
          base_asp = 1.618, base_width = NULL)

```


Supplementary Figure 8
```{r, message=FALSE}

dat_text_mod <- data.frame(
  label = c("*R*<sup>2</sup> = 0.886", # 0.971 NO CONSTANT, 0.886 WITH CONSTANT
            "*R*<sup>2</sup> = 0.292", # 0.859 NO CONSTANT, 0.292 WITH CONSTANT
            "*R*<sup>2</sup> = 0.0695"),
  parameters = c(1, 2, 3),
  x = c(11.25, 11.25, 11.25),
  y = c(2, 2, 2)
)

all_results %>% 
  filter(
    parameters == "1_mod" | parameters == "2_mod" | parameters == "3_rid"
  ) -> all_results_mod

all_results_mod %>% 
  mutate(parameters = case_when(
    parameters == "1_mod" ~ "1",
    parameters == "2_mod" ~ "2",
    parameters == "3_rid" ~ "3"
  )) -> all_results_mod

ggplot(all_results_mod, 
       aes(x=eff_aa, y=eff_aa_est, color = result)) + 
  geom_point() +
  geom_abline(slope = 1, intercept = 0) + 
  labs(
    x = "*n*<sub>eff</sub> of actual distribution", 
    y = "*n*<sub>eff</sub> of fit distribution",
    color = "Result"
  ) +
  scale_color_manual(
    values = c("#E69F00", "#0072B2")
  ) +
  #geom_richtext(data = dat_text, aes(label = label)) #+
  geom_richtext(
    data = dat_text_mod, 
    aes(x = x, y = y,label = label),
    label.color = NA,
    inherit.aes = FALSE
  ) +
  scale_x_continuous(
    limits = c(0, 13),
    breaks = c(2.5, 5.0, 7.5, 10.0, 12.5),
    expand = c(0, 0)
  ) +
  scale_y_continuous(
    limits = c(0, 13),
    breaks = c(0, 2.5, 5, 7.5, 10, 12.5)
  ) +
  facet_grid(vars(parameters)) + 
  theme_cowplot(12, rel_small = 1) +
  panel_border() +
  background_grid() +
  theme(
    axis.title.x = element_markdown(),
    axis.title.y = element_markdown(),
    strip.background = element_rect(fill = "gray90")) -> figs8

figs8

save_plot("figures8.png", figs8, ncol = 1, nrow = 1, base_height = 5,
          base_asp = 1.618, base_width = NULL)


```


Supplementary Figure 6
```{r, message=FALSE}
library(nationalparkcolors)

supp_f6_colors <- park_palette("MtRainier", 3)

chisq_results %>% 
  filter(parameters == 1 | parameters == 2 | parameters == 3) -> chisq_results

chisq_results$parameters <- as.factor(chisq_results$parameters)

chisq_results$parameters <- factor(
  chisq_results$parameters,
  levels = rev(levels(chisq_results$parameters))
)

chisq_results %>% 
  ggplot(aes(x = p.adjusted, y = parameters)) +
  geom_boxplot(aes(fill = parameters), alpha = 0.5) +
  scale_fill_manual(values = supp_f6_colors) +
  labs(
    x = "*p*-value",
    y = "Parameters"
  ) +
  theme(
    legend.position = "none",
    axis.title.x = element_markdown()
  ) +
  panel_border() -> figs6_1

regression_performance %>% 
  pivot_longer(
    cols = c(r2_1, r2_2, r2_3),
    names_to = "parameters",
    values_to = "r2"
  ) -> regression_performance

recode_factor(
  regression_performance$parameters,
  r2_3 = "3",
  r2_2 = "2",
  r2_1 = "1",
  .ordered = TRUE
) -> regression_performance$parameters

regression_performance %>% 
  ggplot(aes(x = parameters, y = r2)) +
  geom_boxplot(aes(fill = parameters), alpha = 0.5) +
  scale_fill_manual(values = supp_f6_colors) +
  coord_flip() +
  labs(y = "*R*<sup>2</sup>") +
  theme(
    axis.title.y = element_blank(),
    legend.position = "none",
    axis.title.x = element_markdown()
  ) +
  panel_border() -> figs6_2

plot_grid(figs6_1, figs6_2, ncol = 2,
          align = "h", axis = "bt", 
          labels = c("A", "B")) -> figs6

figs6

save_plot("figures6.png", figs6, ncol = 1, nrow = 1, base_height = 5,
          base_asp = 1.618, base_width = NULL)


```


Visualize different fits at individual sites
```{r, message=FALSE}
# # visualize how each fit works at any given site
# # 16 should be fit easily
# # 44, 90, 15, 138
# 
# ordered_count %>%
#   filter(site == 16) %>%
#   ggplot(aes(x = k)) +
#   geom_bar(aes(y = count), fill = "grey", stat='identity') +
#   geom_line(aes(y = est_count_1), color = "black", size = 1) +
#   geom_line(aes(y = est_count_2), color = "blue", size = 1) +
#   geom_line(aes(y = est_count_3), color = "violetred", size = 1) +
#   geom_line(aes(y = est_count_1_mod), color = "purple", size = 1) +
#   geom_line(aes(y = est_count_2_mod), color = "green", size = 1) +
#   geom_line(aes(y = est_count_3_rid), color = "orange", size = 1) +
#   # scale_color_manual(
#   #   name = "distribution",
#   #   values = c(
#   #     "grey" = "grey", "black" = "black", "blue" = "blue", "violetred" = "violetred",
#   #     "purple" = "purple", "green" = "green", "orange" = "orange"
#   #   ),
#   #   labels = c("actual", "1 param", "2 param", "3 param", "1 w constant", "2 w constant", "3 ridge")
#   # ) +
#   labs(x = "Amino acids ranked, *i*", y = "Count") +
#   theme(
#     legend.position = "right",
#     axis.title.x = element_markdown()
#   )


```


