---
title: "Fitting parameters"
author: "Mackenzie Johnson"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Varying the number of parameters used for fitting amino acid distributions

This is the RMarkdown document used to produce Figure 6, Supplementary Figure 5, and ###. It compares results for fitting distributions with a different number of parameters.

Load necessary packages
```{r, message=FALSE}
library(tidyverse)
library(Biostrings)
library(stringr)
library(cowplot)
library(broom)
library(ggtext)
theme_set(theme_cowplot())

```


Read in and format data
```{r, message=FALSE}
# read in all .csv files in /data/real/
empirical_files <- dir(
  "data/real", 
  pattern = "*_Aligned_Sequences.fasta", 
  full.names = T
)

# import alignment
import_alignment <- function(filename) {
  
  fasta_file <- readAAStringSet(filename, format = "fasta")
  seq_name = names(fasta_file)
  aa_seq = paste(fasta_file)
  seq_num = seq(1:length(seq_name))
  real_align <- data.frame(seq_num, aa_seq, stringsAsFactors = F)
  
}

# change sequences from strings into character vectors
aa_string_to_vect <- function(real_align) {

  as.data.frame(strsplit(real_align$aa_seq, split = ""), 
                row.names = NULL) -> int_df
  as.data.frame(t(int_df), row.names = NULL) -> int_df

}

# tidy data frame
tidy_df <- function(int_df) {
  int_df %>%
    mutate(num_seq = seq(1:nrow(int_df))) %>% 
    pivot_longer(
      -num_seq, 
      names_to = "site_raw", 
      values_to = "aa"
    ) -> int_df
  int_df %>% 
    extract(
      site_raw, 
      "site", 
      "V(.*)", 
      remove = TRUE
    ) -> int_df
}

#import_alignment(empirical_files[1]) %>% View()
tibble(filename = empirical_files[1]) %>% 
  extract(filename, "protein", "data/real/(.*)_A", remove = FALSE) %>% 
  mutate(
    raw_data = map(filename, import_alignment),
    long_data = map(raw_data, aa_string_to_vect),
    tidy_data = map(long_data, tidy_df)
  ) %>% 
  select(-filename, -raw_data, -long_data) %>% 
  unnest(cols = tidy_data) -> alignment

# replace undefined aas
alignment$site <- as.numeric(alignment$site)

levels(alignment$aa) # "X" = any aa, "B" = N or D, "Z" = Q or D

alignment$aa <- recode(alignment$aa, X = "-", B = "N", Z = "Q")

levels(alignment$aa)

alignment$aa <- as.character(alignment$aa)

```


Actual distribution
```{r, message=FALSE}
# determine the number of each AA at each site
count_aa <- function(protein_alignment) {
  
  protein_alignment %>%
    group_by(site, aa) %>% 
    summarize(count = n())

  }

alignment %>%
  nest(data = c(num_seq, site, aa)) %>% 
  mutate(counts = map(data, count_aa)) %>%
  select(-data) %>% 
  unnest(cols = counts) -> site_aa_count

empty_counts <- tibble(aa = c('A', 'C', 'D', 'E', 'F', 'G', 'H', 
                              'I', 'K', 'L', 'M', 'N', 'P', 'Q', 
                              'R', 'S', 'T', 'V', 'W', 'Y'), 
                           count = rep(0, 20))


# 1. First write a function that can take a data frame with amino acids at one site, and return a data frame with counts, including zeros.
count_zero_aa_site <- function(site_df) {
  
  site_df %>% 
    group_by(aa) %>% 
    summarize(count = n()) -> counts
  full_join(counts, anti_join(empty_counts, counts, by = 'aa'))

  }

# 2. Function to call that function for each protein.
count_zero_aa_protein <- function(protein_df) {
  
  protein_df %>%
    nest(data = -site) %>%
    mutate(counts = map(data, count_zero_aa_site)) %>%
    select(-data) %>%
    unnest(cols = counts)

}

# 3. Apply to all proteins
alignment %>%
  nest(data = c(num_seq, site, aa)) %>%
  mutate(all_counts = map(data, count_zero_aa_protein)) %>%
  select(-data) %>%
  unnest(cols = all_counts) -> all_count


```


Estimated distribution
```{r, message=FALSE}
# USING ONLY OBSERVED AA
# order aa by relative frequency
order_aa <- function(protein_df) {
  
  protein_df %>%
    group_by(site) %>% 
    arrange(desc(count), .by_group = T)
  
}

all_count %>%
  nest(data = c(site, aa, count)) %>% 
  mutate(ordered_count = map(data, order_aa)) %>%
  select(-data) %>% 
  unnest(cols = ordered_count) -> ordered_count

# EXCLUDE all '-' entries/gaps in alignment
ordered_count %>%
  filter(aa != "-") -> no_gaps

# make the categorical variable (aa) numerical (k) by numbering aa's 1-20 in order of freq
number_aa <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(k = as.numeric(0:19))
  
}

no_gaps %>% 
  nest(data = c(site, aa, count)) %>% 
  mutate(add_k = map(data, number_aa)) %>% 
  select(-data) %>% 
  unnest(cols = add_k) -> ordered_count

# remove aa with zero values
ordered_count %>%
  filter(count > 0) -> observed_aa

# count relative to most frequent - rescale to 1 for highly conserved sites
add_rel_count <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(rel_count = count/max(count))
  
}

observed_aa %>%
  nest(data = c(site, aa, count, k)) %>% 
  mutate(rel_count = map(data, add_rel_count)) %>% 
  select(-data) %>% 
  unnest(cols = rel_count) -> observed_aa

# log-transformed data
add_ln_count <- function(protein_df) {
  
  protein_df %>% 
    mutate(ln_count = log(rel_count))
  
}

observed_aa %>%
  nest(data = c(site, aa, count, k, rel_count)) %>% 
  mutate(log_count = map(data, add_ln_count)) %>% 
  select(-data) %>% 
  unnest(log_count) -> observed_aa


```


Add constant to all
```{r}
# add a small constant to include aa with zero values
ordered_count %>%
  mutate(count = count + 0.000001) -> observed_aa_mod
ordered_count %>%
  mutate(count_mod = count + 0.000001) -> ordered_count_mod

# count relative to most frequent - rescale to 1 for highly conserved sites
observed_aa_mod %>%
  group_by(site) %>%
  mutate(rel_count = count/max(count)) -> observed_aa_mod

# log-transformed data
observed_aa_mod %>%
  mutate(ln_count = log(rel_count)) -> observed_aa_mod

# get rid of old data sets
rm(alignment, all_count, empty_counts, no_gaps, site_aa_count)

```


1 parameter fit - without constant
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# set intercept to 0, only fitting 1 parameter (slope)

fit_lm_protein_1 <- function(protein_df) {
  
  protein_df %>% 
    nest(data = c(aa, count, k, rel_count, ln_count)) %>% 
    mutate(
      fit = map(data, ~ lm(ln_count ~ 0 + k, data = .)),
      slope_1 = map_dbl(fit, ~ (.)$coefficients[1])
    ) %>% 
    select(-data, -fit)
  
}

observed_aa %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_1)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_1

```


1 parameter fit - with constant
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# set intercept to 0, only fitting 1 parameter (slope)

observed_aa_mod %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_1)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_mod_1

```


2 parameters fit - without constant
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# fitting 2 parameters: slope and intercept

fit_lm_protein_2 <- function(protein_df) {
  
  protein_df %>% 
    nest(data = c(aa, count, k, rel_count, ln_count)) %>% 
    mutate(
      fit = map(data, ~ lm(ln_count ~ k, data = .)),
      intercept_2 = map_dbl(fit, ~ (.)$coefficients[1]),
      slope_2 = map_dbl(fit, ~ (.)$coefficients[2])
    ) %>% 
    select(-data, -fit)
  
}

observed_aa %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_2)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_2

```


2 parameters fit - with constant
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# fitting 2 parameters: slope and intercept

observed_aa_mod %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_2)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_mod_2

```


3 parameters fit - with constant
```{r, message=FALSE}
# fit quadratic model to normalized, log-transformed data
# fitting 3 parameter: slope, slope^2, intercept
# use non-linear model, fit to all aa (add small value to zero counts)

observed_aa_mod %>%
  nest(data = -site) %>%
  mutate(
    fit = map(data, ~ lm(ln_count ~ k + I(k^2), data = .)),
    intercept_3 = map_dbl(fit, ~ (.)$coefficients[1]),
    slope_3 = map_dbl(fit, ~ (.)$coefficients[2]),
    sqr_slp_3 = map_dbl(fit, ~ (.)$coefficients[3])
  ) %>%
  select(-data, -fit) -> observed_fits_3


```


3 parameter fit - ridge regression
```{r}
# # use ridge regression instead
# #install.packages("glmnet", repos = "https://cran.us.r-project.org")
# library(glmnet)
# 
# # example of use and finding lambda
# data(QuickStartExample)
# fit = glmnet(x, y)
# plot(fit)
# cvfit = cv.glmnet(x, y)
# plot(cvfit)
# cvfit$lambda.min
# 
# # figure out a good lambda value to use
# grid <- 10^seq(10, -2, length = 100)
# 
# observed_aa_mod %>% 
#   filter(site == 1) -> one_site_example
# 
# x_site <- tibble(k = one_site_example$k) %>% 
#   mutate(k2 = k*k)
# x_site <- as.matrix(x_site)
# y_site <- one_site_example$ln_count
# 
# fit <- glmnet(x_site, y_site)
# plot(fit)
# 
# fit = glmnet(
#   x_site, 
#   y_site, 
#   alpha = 0, # for ridge regression
#   lambda = grid
# )
# 
# cvfit <- cv.glmnet(x_site, y_site)
# cvfit$lambda.min
# 
# # lambda value across sites - use one lambda values
# # some lambda value should recreate what you'd get from other model
# # relationship between optimal lambda and neff?
# 
# plot(fit)
# 
# # apply to all sites
# observed_aa_mod %>% 
#   nest(data = -site) %>%
#   mutate(
#     fit = map(data, ~ lm(ln_count ~ k + I(k^2), data = .)),
#     intercept_3 = map_dbl(fit, ~ (.)$coefficients[1]),
#     slope_3 = map_dbl(fit, ~ (.)$coefficients[2]),
#     sqr_slp_3 = map_dbl(fit, ~ (.)$coefficients[3])
#   ) %>% 
#   select(-data, -fit) -> observed_fits_3
# 

```


Get estimate counts 
```{r, message=FALSE}
# combine dataframes of fits (no constant for 1 and 2 parameter)
observed_fits_1 %>%
  full_join(., observed_fits_2, by = c("protein", "site")) %>% 
  full_join(., observed_fits_3, by = "site") -> observed_fits

# combine dataframes of fits (add constant to all)
observed_fits_mod_1 %>%
  full_join(., observed_fits_mod_2, by = c("protein", "site")) %>% 
  full_join(., observed_fits_3, by = "site") -> observed_fits_mod

# get rid of old subsets
rm(observed_fits_1, observed_fits_2, observed_fits_mod_1, 
   observed_fits_mod_2, observed_fits_3)

# added rel_count and ln_count to all aa
ordered_count %>%
  nest(data = c(site, aa, count, k)) %>% 
  mutate(rel_count = map(data, add_rel_count)) %>% 
  select(-data) %>% 
  unnest(cols = rel_count) -> ordered_count

ordered_count %>%
  nest(data = c(site, aa, count, k, rel_count)) %>% 
  mutate(ln_count = map(data, add_ln_count)) %>% 
  select(-data) %>% 
  unnest(cols = ln_count) -> ordered_count

# estimate counts with parameters from fit - no constant for 1 and 2 param
ordered_count %>% 
  left_join(observed_fits, by = c("protein", "site")) %>%
  mutate(
    est_dist_1 = slope_1*exp(k*slope_1),
    est_dist_2 = slope_2*exp(k*slope_2+intercept_2),
    est_dist_3 = slope_3*exp(k*slope_3 + k*k*sqr_slp_3 + 
                                    intercept_3)
  ) -> ordered_count

# estimate counts with parameters from fit - constant for all
colnames(observed_fits_mod)[3] <- "slope_1_mod"
colnames(observed_fits_mod)[4] <- "intercept_2_mod"
colnames(observed_fits_mod)[5] <- "slope_2_mod"

observed_fits_mod <- observed_fits_mod[1:5]

ordered_count %>% 
  left_join(
    observed_fits_mod, 
    by = c("protein", "site")
  ) %>%
  mutate(
    est_dist_1_mod = slope_1_mod*exp(k*slope_1_mod),
    est_dist_2_mod = slope_2_mod*exp(k*slope_2_mod+intercept_2_mod)
  ) -> ordered_count


# rescale to compare w raw count
rescale_est_dist <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(
      est_rel_1 = est_dist_1/sum(est_dist_1),
      est_rel_2 = est_dist_2/sum(est_dist_2),
      est_rel_3 = est_dist_3/sum(est_dist_3),
      est_rel_1_mod = est_dist_1_mod/sum(est_dist_1_mod),
      est_rel_2_mod = est_dist_2_mod/sum(est_dist_2_mod)
    ) %>% 
    mutate(
      est_count_1 = sum(count)*est_rel_1,
      est_count_2 = sum(count)*est_rel_2,
      est_count_3 = sum(count)*est_rel_3,
      est_count_1_mod = sum(count)*est_rel_1_mod,
      est_count_2_mod = sum(count)*est_rel_2_mod
    )
  
}

ordered_count %>%
  nest(data = c(site, aa, count, k, rel_count, 
                ln_count, slope_1, intercept_2,
                slope_2, intercept_3, slope_3,
                sqr_slp_3, est_dist_1, est_dist_2, 
                est_dist_3, est_dist_1_mod, 
                est_dist_2_mod)) %>% 
  mutate(rescaled = map(data, rescale_est_dist)) %>% 
  select(-data) %>% 
  unnest(cols = rescaled) -> ordered_count

# remove duplicate information
rm(observed_fits, observed_fits_mod)

```



Chi-squared test: actual vs estimated
```{r, message=FALSE}
# use raw count
ordered_count %>% 
  nest(data = -site) %>%
  ungroup() %>%
  mutate(
    chisq_1 = map(data, ~ sum(((.$count - .$est_count_1)^2)/.$est_count_1)),
    chisq_2 = map(data, ~ sum(((.$count - .$est_count_2)^2)/.$est_count_2)),
    chisq_3 = map(data, ~ sum(((.$count - .$est_count_3)^2)/.$est_count_3)),
    chisq_1_mod = map(data, ~ sum(((.$count - .$est_count_1_mod)^2)/.$est_count_1_mod)),
    chisq_2_mod = map(data, ~ sum(((.$count - .$est_count_2_mod)^2)/.$est_count_2_mod)),
    p_value_1 = map_dbl(chisq_1, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_2 = map_dbl(chisq_2, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_3 = map_dbl(chisq_3, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_1_mod = map_dbl(chisq_1_mod, ~ pchisq(., 18, lower.tail=FALSE)),
    p_value_2_mod = map_dbl(chisq_2_mod, ~ pchisq(., 18, lower.tail=FALSE)),
  ) %>% 
  select(-data) %>%
  mutate(
    p.adjusted_1 = p.adjust(p_value_1, method= "fdr"),
    p.adjusted_2 = p.adjust(p_value_2, method= "fdr"),
    p.adjusted_3 = p.adjust(p_value_3, method= "fdr"),
    p.adjusted_1_mod = p.adjust(p_value_1_mod, method= "fdr"),
    p.adjusted_2_mod = p.adjust(p_value_2_mod, method= "fdr"),
  ) -> chisq_results

chisq_results %>% 
  select(
    site, p.adjusted_1, p.adjusted_2, p.adjusted_3,
    p.adjusted_1_mod, p.adjusted_2_mod
  ) -> chisq_results

pivot_longer(
  chisq_results, 
  -site, 
  names_to = "parameters_raw", 
  values_to = "p.adjusted"
) %>% 
  extract(
    parameters_raw, 
    "parameters", 
    "p.adjusted_(.*)", 
    remove = T
  ) -> chisq_results

chisq_results %>%
  mutate(result = "pass") -> chisq_results

# REMOVE conserved sites with a single aa
chisq_results %>% 
  filter(p.adjusted != "NaN") -> chisq_results

chisq_results$result[chisq_results$p.adjusted < 0.05] <- "fail"

```


Chi-squared results based on number of parameters used
```{r, message=FALSE}
# 133 sites; 17 fail (12.8%) - NO CONSTANT
chisq_results %>% 
  filter(parameters == 1 & result == "fail") %>% nrow()

# 153 sites; 133 fail (86.9%) - WITH CONSTANT
chisq_results %>% 
  filter(parameters == "1_mod" & result == "fail") %>% nrow()

# 133 sites; 34 fail (25.6%) - NO CONSTANT
chisq_results %>% 
  filter(parameters == 2 & result == "fail") %>% nrow()

# 153 sites; 135 fail (88.2%) - WITH CONSTANT
chisq_results %>% 
  filter(parameters == "2_mod" & result == "fail") %>% nrow()

# 153 sites; 129 fail (84.3%) - WITH CONSTANT
chisq_results %>% 
  filter(parameters == 3 & result == "fail") %>% nrow()

```


Effective number of amino acids
```{r, message=FALSE}
# eff aa for actual distribution
observed_aa %>%
  group_by(site) %>%
  mutate(frequency = (count)/sum(count)) -> site_aa_freq

site_aa_freq %>% 
  mutate(flnf = frequency*log(frequency)) %>%
  group_by(site) %>% 
  summarize(
    entropy = -sum(flnf),
    eff_aa = exp(entropy),
    n = length(unique(aa))
  ) -> eff_aa_all

# eff aa for estimated distributions
ordered_count %>% 
  group_by(site) %>%
  mutate(
    freq_1 = (est_count_1)/sum(est_count_1),
    freq_2 = (est_count_2)/sum(est_count_2),
    freq_3 = (est_count_3)/sum(est_count_3),
    freq_1_mod = (est_count_1_mod)/sum(est_count_1_mod),
    freq_2_mod = (est_count_2_mod)/sum(est_count_2_mod),
  ) %>% 
  select(
    site, aa, k, freq_1, freq_2, freq_3,
    freq_1_mod, freq_2_mod
  ) -> site_aa_freq_2

site_aa_freq_2 %>% 
  mutate(
    flnf_1 = freq_1*log(freq_1),
    flnf_2 = freq_2*log(freq_2),
    flnf_3 = freq_3*log(freq_3),
    flnf_1_mod = freq_1_mod*log(freq_1_mod),
    flnf_2_mod = freq_2_mod*log(freq_2_mod)
  ) %>%
  group_by(site) %>%
  summarize(
    entr_1 = -sum(flnf_1),
    entr_2 = -sum(flnf_2),
    entr_3 = -sum(flnf_3),
    entr_1_mod = -sum(flnf_1_mod),
    entr_2_mod = -sum(flnf_2_mod),
    eff_aa_est_1 = exp(entr_1),
    eff_aa_est_2 = exp(entr_2),
    eff_aa_est_3 = exp(entr_3),
    eff_aa_est_1_mod = exp(entr_1_mod),
    eff_aa_est_2_mod = exp(entr_2_mod)
  ) -> eff_aa_all_est

eff_aa_all_est %>% 
  pivot_longer(
    c(entr_1, entr_2, entr_3, entr_1_mod, entr_2_mod), 
    names_to = "parameters",
    values_to = "entropy_est"
  ) %>% 
  extract(
    parameters, 
    "parameters", 
    "entr_(.*)", 
    remove = T
  ) %>% 
  select(site, parameters, entropy_est) -> sub1
  
eff_aa_all_est %>% 
  pivot_longer(
    c(
      eff_aa_est_1, eff_aa_est_2, eff_aa_est_3, 
      eff_aa_est_1_mod, eff_aa_est_2_mod
    ),
    names_to = "parameter",
    values_to = "eff_aa_est"
  ) %>% 
  extract(
    parameter, 
    "parameters", 
    "eff_aa_est_(.*)", 
    remove = T
  ) %>%
  select(site, parameters, eff_aa_est) -> sub2

full_join(sub1, sub2, by = c("site", "parameters")) -> eff_aa_est

left_join(eff_aa_all, eff_aa_est, by = "site") -> compare_eff_aa

rm(sub1, sub2, site_aa_freq, site_aa_freq_2, 
   observed_aa, observed_aa_mod, ordered_count, 
   ordered_count_mod, eff_aa_all_est, eff_aa_all, 
   eff_aa_est)

```


Find $R^2$ values for each parameter
```{r,message=FALSE}

rsq <- function (x, y) { cor(x, y) ^ 2 }

compare_eff_aa %>% 
  filter(parameters == 1 & !is.na(eff_aa_est)) -> compare_p1

rsq(compare_p1$eff_aa, compare_p1$eff_aa_est) # 0.9706805 - NO CONSTANT 

compare_eff_aa %>% 
  filter(parameters == "1_mod" & !is.na(eff_aa_est)) -> compare_p1_mod

rsq(compare_p1_mod$eff_aa, compare_p1_mod$eff_aa_est) # 0.7409036 - WITH CONSTANT

compare_eff_aa %>% 
  filter(parameters == 2 & !is.na(eff_aa_est)) -> compare_p2

rsq(compare_p2$eff_aa, compare_p2$eff_aa_est) # 0.8589123 - NO CONSTANT

compare_eff_aa %>% 
  filter(parameters == "2_mod" & !is.na(eff_aa_est)) -> compare_p2_mod

rsq(compare_p2_mod$eff_aa, compare_p2_mod$eff_aa_est) # 0.1839351 - WITH CONSTANT

compare_eff_aa %>% 
  filter(parameters == 3 & !is.na(eff_aa_est)) -> compare_p3

rsq(compare_p3$eff_aa, compare_p3$eff_aa_est) # 0.718942 - WITH CONSTANT

rm(compare_p1, compare_p1_mod, compare_p2, compare_p2_mod, compare_p3)

```


Figure 6
```{r, message=FALSE}

signif(0.9706805, digits = 3)
signif(0.8589123, digits = 3)
signif(0.718942, digits = 3)

dat_text <- data.frame(
  label = c("*R*<sup>2</sup> = 0.971", # 0.971 NO CONSTANT, 0.741 WITH CONSTANT
            "*R*<sup>2</sup> = 0.859", # 0.859 NO CONSTANT, 0.184 WITH CONSTANT
            "*R*<sup>2</sup> = 0.719"),
  parameters = c(1, 2, 3),
  x = c(11.25, 11.25, 11.25),
  y = c(2, 2, 2)
)


left_join(compare_eff_aa, chisq_results) %>% 
  filter(!is.na(result)) -> all_results

all_results %>% 
  filter(
    parameters == 1 | parameters == 2 | parameters == 3
  ) -> all_results_original

ggplot(all_results_original, 
       aes(x=eff_aa, y=eff_aa_est, color = result)) + 
  geom_point() +
  geom_abline(slope = 1, intercept = 0) + 
  labs(
    x = "*n*<sub>eff</sub> of actual distribution", 
    y = "*n*<sub>eff</sub> of fit distribution",
    color = "Result"
  ) +
  scale_color_manual(
    values = c("#E69F00", "#0072B2")
  ) +
  #geom_richtext(data = dat_text, aes(label = label)) #+
  geom_richtext(
    data = dat_text, 
    aes(x = x, y = y,label = label),
    label.color = NA,
    inherit.aes = FALSE
  ) +
  facet_grid(vars(parameters)) + 
  theme_cowplot(12, rel_small = 1) +
  panel_border() +
  background_grid() +
  theme(
    axis.title.x = element_markdown(),
    axis.title.y = element_markdown(),
    strip.background = element_rect(fill = "gray90")) -> fig6

fig6

save_plot("figure6.png", fig6, ncol = 1, nrow = 1, base_height = 5,
          base_asp = 1.618, base_width = NULL)

```


Supplementary Figure 5
```{r}
dat_text_mod <- data.frame(
  label = c("*R*<sup>2</sup> = 0.741", # 0.971 NO CONSTANT, 0.741 WITH CONSTANT
            "*R*<sup>2</sup> = 0.184", # 0.859 NO CONSTANT, 0.184 WITH CONSTANT
            "*R*<sup>2</sup> = 0.719"),
  parameters = c(1, 2, 3),
  x = c(11.25, 11.25, 11.25),
  y = c(2, 2, 2)
)

all_results %>% 
  filter(
    parameters == "1_mod" | parameters == "2_mod" | parameters == 3
  ) -> all_results_mod

all_results_mod %>% 
  mutate(parameters = case_when(
    parameters == "1_mod" ~ "1",
    parameters == "2_mod" ~ "2",
    TRUE ~ "3"
  )) -> all_results_mod

ggplot(all_results_mod, 
       aes(x=eff_aa, y=eff_aa_est, color = result)) + 
  geom_point() +
  geom_abline(slope = 1, intercept = 0) + 
  labs(
    x = "*n*<sub>eff</sub> of actual distribution", 
    y = "*n*<sub>eff</sub> of fit distribution",
    color = "Result"
  ) +
  scale_color_manual(
    values = c("#E69F00", "#0072B2")
  ) +
  #geom_richtext(data = dat_text, aes(label = label)) #+
  geom_richtext(
    data = dat_text_mod, 
    aes(x = x, y = y,label = label),
    label.color = NA,
    inherit.aes = FALSE
  ) +
  facet_grid(vars(parameters)) + 
  theme_cowplot(12, rel_small = 1) +
  panel_border() +
  background_grid() +
  theme(
    axis.title.x = element_markdown(),
    axis.title.y = element_markdown(),
    strip.background = element_rect(fill = "gray90")) -> figs5

figs5

save_plot("figures5.png", figs5, ncol = 1, nrow = 1, base_height = 5,
          base_asp = 1.618, base_width = NULL)


```

