---
title: "Fitting parameters"
author: "Mackenzie Johnson"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Varying the number of parameters used for fitting amino acid distributions

This is the RMarkdown document used to produce Supplementary Figure 5. It compares results for fitting distributions with a different number of parameters for all empirical MSAs.

Load necessary packages
```{r, message=FALSE}
library(tidyverse)
library(Biostrings)
library(stringr)
library(cowplot)
library(broom)
library(ggtext)
theme_set(theme_cowplot())

```


Read in and format data
```{r, message=FALSE}
# read in all .csv files in /data/real/
empirical_files <- dir(
  "data/real", 
  pattern = "*_Aligned_Sequences.fasta", 
  full.names = T
)

# import alignment
import_alignment <- function(filename) {
  
  fasta_file <- readAAStringSet(filename, format = "fasta")
  seq_name = names(fasta_file)
  aa_seq = paste(fasta_file)
  seq_num = seq(1:length(seq_name))
  real_align <- data.frame(seq_num, aa_seq, stringsAsFactors = F)
  
}

# change sequences from strings into character vectors
aa_string_to_vect <- function(real_align) {

  as.data.frame(strsplit(real_align$aa_seq, split = ""), 
                row.names = NULL) -> int_df
  as.data.frame(t(int_df), row.names = NULL) -> int_df

}

# tidy data frame
tidy_df <- function(int_df) {
  int_df %>%
    mutate(num_seq = seq(1:nrow(int_df))) %>% 
    pivot_longer(-num_seq, 
                 names_to = "site_raw", 
                 values_to = "aa") -> int_df
  int_df %>% 
    extract(site_raw, "site", "V(.*)", remove = TRUE) -> int_df
}

#import_alignment(empirical_files[1]) %>% View()
tibble(filename = empirical_files) %>% 
  extract(
    filename, 
    "protein", 
    "data/real/(.*)_A", 
    remove = FALSE
  ) %>% 
  mutate(
    raw_data = map(filename, import_alignment),
    long_data = map(raw_data, aa_string_to_vect),
    tidy_data = map(long_data, tidy_df)
  ) %>% 
  select(-filename, -raw_data, -long_data) %>% 
  unnest(cols = tidy_data) -> alignments

# replace undefined aas
alignments$site <- as.numeric(alignments$site)

levels(alignments$aa) # "X" = any aa, "B" = N or D, "Z" = Q or D

alignments$aa <- recode(alignments$aa, X = "-", B = "N", Z = "Q")

levels(alignments$aa)

alignments$aa <- as.character(alignments$aa)

```


Actual distribution
```{r, message=FALSE}
# determine the number of each AA at each site
count_aa <- function(protein_alignment) {
  
  protein_alignment %>%
    group_by(site, aa) %>% 
    summarize(count = n())

  }

alignments %>%
  nest(data = c(num_seq, site, aa)) %>% 
  mutate(counts = map(data, count_aa)) %>%
  select(-data) %>% 
  unnest(cols = counts) -> site_aa_count

empty_counts <- tibble(aa = c('A', 'C', 'D', 'E', 'F', 'G', 'H', 
                              'I', 'K', 'L', 'M', 'N', 'P', 'Q', 
                              'R', 'S', 'T', 'V', 'W', 'Y'), 
                           count = rep(0, 20))


# 1. First write a function that can take a data frame with amino acids at one site, and return a data frame with counts, including zeros.
count_zero_aa_site <- function(site_df) {
  
  site_df %>% 
    group_by(aa) %>% 
    summarize(count = n()) -> counts
  full_join(counts, anti_join(empty_counts, counts, by = 'aa'))

  }

# 2. Function to call that function for each protein.
count_zero_aa_protein <- function(protein_df) {
  
  protein_df %>%
    nest(data = -site) %>%
    mutate(counts = map(data, count_zero_aa_site)) %>%
    select(-data) %>%
    unnest(cols = counts)

}

# 3. Apply to all proteins
alignments %>%
  nest(data = c(num_seq, site, aa)) %>%
  mutate(all_counts = map(data, count_zero_aa_protein)) %>%
  select(-data) %>%
  unnest(cols = all_counts) -> all_count


```


Estimated distribution
```{r, message=FALSE}
# USING ONLY OBSERVED AA
# order aa by relative frequency
order_aa <- function(protein_df) {
  
  protein_df %>%
    group_by(site) %>% 
    arrange(desc(count), .by_group = T)
  
}

all_count %>%
  nest(data = c(site, aa, count)) %>% 
  mutate(ordered_count = map(data, order_aa)) %>%
  select(-data) %>% 
  unnest(cols = ordered_count) -> ordered_count

# EXCLUDE all '-' entries/gaps in alignment
ordered_count %>%
  filter(aa != "-") -> no_gaps

# make the categorical variable (aa) numerical (k) by numbering aa's 1-20 in order of freq
number_aa <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(k = as.numeric(0:19))
  
}

no_gaps %>% 
  nest(data = c(site, aa, count)) %>% 
  mutate(add_k = map(data, number_aa)) %>% 
  select(-data) %>% 
  unnest(cols = add_k) -> ordered_count

# remove aa with zero values
ordered_count %>%
  filter(count > 0) -> observed_aa

# count relative to most frequent - rescale to 1 for highly conserved sites
add_rel_count <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(rel_count = count/max(count))
  
}

observed_aa %>%
  nest(data = c(site, aa, count, k)) %>% 
  mutate(rel_count = map(data, add_rel_count)) %>% 
  select(-data) %>% 
  unnest(cols = rel_count) -> observed_aa

# log-transformed data
add_ln_count <- function(protein_df) {
  
  protein_df %>% 
    mutate(ln_count = log(rel_count))
  
}

observed_aa %>%
  nest(data = c(site, aa, count, k, rel_count)) %>% 
  mutate(log_count = map(data, add_ln_count)) %>% 
  select(-data) %>% 
  unnest(log_count) -> observed_aa

```


1 parameter fit
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# set intercept to 0, only fitting 1 parameter (slope)

fit_lm_protein_1 <- function(protein_df) {
  
  protein_df %>% 
    nest(data = c(aa, count, k, rel_count, ln_count)) %>% 
    mutate(fit = map(data, ~ lm(ln_count ~ 0 + k, data = .)),
           slope_1 = map_dbl(fit, ~ (.)$coefficients[1])) %>% 
    select(-data, -fit)
  
}

observed_aa %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_1)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_1

```



2 parameters fit
```{r, message=FALSE}
# fit linear model to normalized, log-transformed data
# fitting 2 parameters: slope and intercept

fit_lm_protein_2 <- function(protein_df) {
  
  protein_df %>% 
    nest(data = c(aa, count, k, rel_count, ln_count)) %>% 
    mutate(
      fit = map(data, ~ lm(ln_count ~ k, data = .)),
      intercept_2 = map_dbl(fit, ~ (.)$coefficients[1]),
      slope_2 = map_dbl(fit, ~ (.)$coefficients[2])
    ) %>% 
    select(-data, -fit)
  
}

observed_aa %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_2)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_2

```


3 parameters fit
```{r, message=FALSE}
# fit quadratic model to normalized, log-transformed data
# fitting 3 parameter: slope, slope^2, intercept
# use non-linear model, fit to all aa (add small value to zero counts)

# add a small constant to include aa with zero values
ordered_count %>%
  mutate(count = count + 0.01) -> observed_aa_mod
  #mutate(count = count + 0.000001) -> observed_aa_mod
ordered_count %>%
  mutate(count_mod = count + 0.01) -> ordered_count_mod
  #mutate(count_mod = count + 0.000001) -> ordered_count_mod

# count relative to most frequent - rescale to 1 for highly conserved sites
observed_aa_mod %>%
  group_by(site) %>%
  mutate(rel_count = count/max(count)) -> observed_aa_mod

# log-transformed data
observed_aa_mod %>%
  mutate(ln_count = log(rel_count)) -> observed_aa_mod 

fit_lm_protein_3 <- function(protein_df) {
  
  protein_df %>% 
    nest(data = c(aa, count, k, rel_count, ln_count)) %>% 
    mutate(
      fit = map(data, ~ lm(ln_count ~ k + I(k^2), data = .)),
      intercept_3 = map_dbl(fit, ~ (.)$coefficients[1]),
      slope_3 = map_dbl(fit, ~ (.)$coefficients[2]),
      sqr_slp_3 = map_dbl(fit, ~ (.)$coefficients[3])
    ) %>% 
    select(-data, -fit)
  
}

observed_aa_mod %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count)) %>% 
  mutate(fits = map(data, fit_lm_protein_3)) %>% 
  select(-data) %>% 
  unnest(cols = fits) -> observed_fits_3

```


Get estimate counts
```{r, message=FALSE}
# combine dataframes of fits
observed_fits_1 %>%
  full_join(., observed_fits_2, by = c("protein", "site")) %>% 
  full_join(., observed_fits_3, by = c("protein", "site")) -> observed_fits

# added rel_count and ln_count to all aa
ordered_count %>%
  nest(data = c(site, aa, count, k)) %>% 
  mutate(rel_count = map(data, add_rel_count)) %>% 
  select(-data) %>% 
  unnest(cols = rel_count) -> ordered_count

ordered_count %>%
  nest(data = c(site, aa, count, k, rel_count)) %>% 
  mutate(ln_count = map(data, add_ln_count)) %>% 
  select(-data) %>% 
  unnest(cols = ln_count) -> ordered_count

ordered_count %>% 
  left_join(observed_fits, by = c("protein", "site")) %>%
  mutate(est_dist_1 = slope_1*exp(k*slope_1)) -> ordered_count

ordered_count %>% 
  mutate(
    est_dist_2 = slope_2*exp(k*slope_2+intercept_2),
    est_dist_3 = slope_3*exp(k*slope_3 + k*k*sqr_slp_3 + intercept_3)
  ) -> ordered_count

# rescale to compare w raw count
rescale_est_dist <- function(protein_df) {
  
  protein_df %>% 
    group_by(site) %>% 
    mutate(
      est_rel_1 = est_dist_1/sum(est_dist_1),
      est_rel_2 = est_dist_2/sum(est_dist_2),
      est_rel_3 = est_dist_3/sum(est_dist_3)
    ) %>% 
    mutate(
      est_count_1 = sum(count)*est_rel_1,
      est_count_2 = sum(count)*est_rel_2,
      est_count_3 = sum(count)*est_rel_3
    )
  
}

ordered_count %>%
  nest(data = c(site, aa, count, k, rel_count, 
                ln_count, slope_1, intercept_2,
                slope_2, intercept_3, slope_3,
                sqr_slp_3, est_dist_1, est_dist_2, 
                est_dist_3)) %>% 
  mutate(rescaled = map(data, rescale_est_dist)) %>% 
  select(-data) %>% 
  unnest(cols = rescaled) -> ordered_count

```



Chi-squared test: actual vs estimated
```{r, message=FALSE}
# use raw count
chi_square_protein <- function(protein_df) {
  
  protein_df %>% 
    nest(data = c(aa, count, k, rel_count, ln_count, slope_1,
                  intercept_2, slope_2, intercept_3,
                  slope_3, sqr_slp_3, est_dist_1, est_dist_2,
                  est_dist_3, est_rel_1, est_rel_2, est_rel_3, 
                  est_count_1, est_count_2, est_count_3)) %>% 
    ungroup() %>% 
    mutate(
      chisq_1 = map(data, ~ sum(((.$count - .$est_count_1)^2)/.$est_count_1)),
      chisq_2 = map(data, ~ sum(((.$count - .$est_count_2)^2)/.$est_count_2)),
      chisq_3 = map(data, ~ sum(((.$count - .$est_count_3)^2)/.$est_count_3)),
      p_value_1 = map_dbl(chisq_1, ~ pchisq(., 18, lower.tail=FALSE)),
      p_value_2 = map_dbl(chisq_2, ~ pchisq(., 18, lower.tail=FALSE)),
      p_value_3 = map_dbl(chisq_3, ~ pchisq(., 18, lower.tail=FALSE))
    ) %>% 
    select(-data) %>% 
    mutate(
      p.adjusted_1 = p.adjust(p_value_1, method= "fdr"),
      p.adjusted_2 = p.adjust(p_value_2, method= "fdr"),
      p.adjusted_3 = p.adjust(p_value_3, method= "fdr")
    )
  
}

ordered_count %>% 
  nest(data = c(site, aa, count, k, rel_count, ln_count, slope_1,
                intercept_2, slope_2, intercept_3,
                slope_3, sqr_slp_3, est_dist_1, est_dist_2,
                est_dist_3, est_rel_1, est_rel_2, est_rel_3, 
                est_count_1, est_count_2, est_count_3)) %>% 
  mutate(chisq_results = map(data, chi_square_protein)) %>% 
  select(-data) %>% 
  unnest(chisq_results) -> chisq_results

chisq_results %>% 
  select(protein, site, p.adjusted_1, p.adjusted_2, p.adjusted_3) -> chisq_results

pivot_longer(
  chisq_results,
  c(p.adjusted_1, p.adjusted_2, p.adjusted_3), 
  names_to = "parameters_raw", 
  values_to = "p.adjusted"
) %>% 
  extract(
    parameters_raw, 
    "parameters", 
    "p.adjusted_(.*)", 
    remove = T
  ) -> chisq_results

chisq_results %>%
  mutate(result = "pass") -> chisq_results

# REMOVE conserved sites with a single aa
chisq_results %>% 
  filter(p.adjusted != "NaN") -> chisq_results

chisq_results$result[chisq_results$p.adjusted < 0.05] <- "fail"

```


Chi-squared results based on number of parameters used
```{r, message=FALSE}
# 1834 sites; 254 fail 
chisq_results %>% 
  filter(parameters == 1 & result == "fail") %>% nrow()

# 1834 sites; 486 fail 
chisq_results %>% 
  filter(parameters == 2 & result == "fail") %>% nrow()

# 2058 sites; 1253 fail 
chisq_results %>% 
  filter(parameters == 3 & result == "fail") %>% nrow()

```



Effective number of amino acids
```{r, message=FALSE}
# eff aa for actual distribution
observed_aa %>%
  group_by(protein, site) %>%
  mutate(frequency = (count)/sum(count)) -> site_aa_freq

site_aa_freq %>% 
  mutate(flnf = frequency*log(frequency)) %>%
  group_by(protein, site) %>% 
  summarize(entropy = -sum(flnf),
            eff_aa = exp(entropy),
            n = length(unique(aa))) -> eff_aa_all

# eff aa for estimated distributions
ordered_count %>% 
  group_by(protein, site) %>%
  mutate(
    freq_1 = (est_count_1)/sum(est_count_1),
    freq_2 = (est_count_2)/sum(est_count_2),
    freq_3 = (est_count_3)/sum(est_count_3)
  ) %>% 
  select(site, aa, k, freq_1, freq_2, freq_3) -> site_aa_freq_2
site_aa_freq_2 %>% 
  mutate(
    flnf_1 = freq_1*log(freq_1),
    flnf_2 = freq_2*log(freq_2),
    flnf_3 = freq_3*log(freq_3)
  ) %>%
  group_by(protein, site) %>%
  summarize(
    entr_1 = -sum(flnf_1),
    entr_2 = -sum(flnf_2),
    entr_3 = -sum(flnf_3),
    eff_aa_est_1 = exp(entr_1),
    eff_aa_est_2 = exp(entr_2),
    eff_aa_est_3 = exp(entr_3)
  ) -> eff_aa_all_est

eff_aa_all_est %>% 
  pivot_longer(
    c(entr_1, entr_2, entr_3), 
    names_to = "parameters",
    values_to = "entropy_est"
  ) %>% 
  extract(
    parameters, 
    "parameters", 
    "entr_(.*)", 
    remove = T
  ) %>% 
  select(protein, site, parameters, entropy_est) -> sub1
  
eff_aa_all_est %>% 
  pivot_longer(
    c(eff_aa_est_1, eff_aa_est_2, eff_aa_est_3),
    names_to = "parameter",
    values_to = "eff_aa_est"
  ) %>% 
  extract(
    parameter, 
    "parameters", 
    "eff_aa_est_(.*)", 
    remove = T
  ) %>%
  select(protein, site, parameters, eff_aa_est) -> sub2

full_join(
  sub1, 
  sub2, 
  by = c("protein", "site", "parameters")
) -> eff_aa_est

left_join(
  eff_aa_all, 
  eff_aa_est, 
  by = c("protein", "site")
) -> compare_eff_aa


```


Find $R^2$ values for each parameter
```{r,message=FALSE}

rsq <- function (x, y) { cor(x, y) ^ 2 }

compare_eff_aa %>% 
  filter(parameters == 1 & !is.na(eff_aa_est)) -> compare_p1

compare_p1 %>% 
  group_by(protein) %>% 
  summarise(r2_1 = rsq(eff_aa, eff_aa_est)) -> results_p1

compare_eff_aa %>% 
  filter(parameters == 2 & !is.na(eff_aa_est)) -> compare_p2

compare_p2 %>% 
  group_by(protein) %>% 
  summarise(r2_2 = rsq(eff_aa, eff_aa_est)) -> results_p2

compare_eff_aa %>% 
  filter(parameters == 3 & !is.na(eff_aa_est)) -> compare_p3

compare_p3 %>% 
  group_by(protein) %>% 
  summarise(r2_3 = rsq(eff_aa, eff_aa_est)) -> results_p3

full_join(results_p1, results_p2) %>% 
  full_join(., results_p3) %>% 
  pivot_longer(
    c(r2_1, r2_2, r2_3),
    names_to = "parameters",
    values_to = "r2"
  ) %>% 
  extract(
    parameters, 
    c("trash", "param"), 
    "(r2_)(.*)"
  ) %>% 
  select(-trash) -> results_corr


```


Supplemental Figure 5
```{r, message=FALSE}

ggplot(results_corr, aes(x = param, y = r2)) +
  geom_boxplot(fill = "white", outlier.shape = NA) +
  #geom_jitter(height = 0, width = .1) +
  #geom_point(position = position_jitter(height = 0, width = 0.1, seed = 10)) +
  #geom_line(aes(group = protein), position = position_jitter(height = 0, 
  #                                                           width = 0.1,
  #                                                           seed = 10)) +
  geom_line(aes(group = protein), color = "gray") +
  geom_point() +
  labs(x = "Number of parameters in regression", 
       y = "*R*<sup>2</sup>") +
  theme_cowplot(12, rel_small = 1) +
  theme(
    axis.title.x = element_markdown(),
    axis.title.y = element_markdown()) -> figs5

figs5

save_plot("figures5.png", figs5, ncol = 1, nrow = 1, base_height = 5,
          base_asp = 1.618, base_width = NULL)

```


